VoxelSniperCore/src/main/java/com/github/kevindagame/brush/polymorphic/PolyBrush.kt

```diff
@@ -0,0 +1,212 @@
+    val shapes: MutableList<PolyBrushShape>,
+    val operationType: PolyOperationType,
+    val operation: PolyOperation?,
+) : AbstractBrush() {
+    var properties: MutableList<PolyProperty<*>> = mutableListOf()
+    private var positions: MutableList<BaseLocation> = ArrayList()
+
+    init {
+        this.name = name
+        this.permissionNode = permissionNode
+        val properties = mutableSetOf<PolyPropertiesEnum>()
+        for (shape in shapes) {
+            properties.addAll(shape.parameters.toList())
+        }
+        if (operation != null) {
+            properties.add(PolyPropertiesEnum.EXCLUDEWATER)
+            properties.add(PolyPropertiesEnum.EXCLUDEAIR)
+
+        }
+        properties.add(PolyPropertiesEnum.PERFORMER)
+        for (property in properties) {
+            this.properties.add(property.supplier())
+        }
+    }
+
+    override fun info(vm: VoxelMessage) {
+        vm.brushName(name)
+        vm.size()
+    }
+
+    override fun arrow(v: SnipeData) {
+        positions.clear()
+        doArrow(v)
+    }
+
+    override fun powder(v: SnipeData) {
+        positions.clear()
+        doPowder(v)
+    }
+
+    fun doArrow(v: SnipeData) {
+        executeBrush(v)
+    }
+
+    fun doPowder(v: SnipeData) {
+        executeBrush(v)
+    }
+
+
+    fun executeBrush(v: SnipeData) {
+        this.positions = getPositions(v)
+
+        if (operation != null) {
+            val brushSize = v.brushSize
+            val newMaterials = operation.apply(brushSize, this, excludeAir, excludeWater)
+            for (position in positions) {
+                val material: VoxelMaterial =
+                    newMaterials[position.blockX - targetBlock.x + brushSize][position.blockY - targetBlock.y + brushSize][position.blockZ - targetBlock.z + brushSize]
+                if (!(excludeAir && material.isAir) && !(excludeWater && material === VoxelMaterial.WATER)) {
+                    addOperation(BlockOperation(position, position.block.blockData, material.createBlockData()))
+                }
+            }
+        } else {
+            addOperations(currentPerformer.perform(positions, v))
+        }
+    }
+
+    private fun getPositions(v: SnipeData): MutableList<BaseLocation> {
+        val positions = initPositions(v)
+        val newPositions = mutableListOf<BaseLocation>()
+        val radiusSquared = (v.brushSize + if (smooth) SMOOTH_CIRCLE_VALUE else VOXEL_CIRCLE_VALUE).pow(2)
+        val center = targetBlock
+
+        for (position in positions) {
+            for (shape in shapes) {
+                if (shape.apply(v, position, radiusSquared)) {
+                    newPositions.add(
+                        BaseLocation(
+                            world,
+                            (center.x + position.dx).toDouble(),
+                            (center.y + position.dy).toDouble(),
+                            (center.z + position.dz).toDouble()
+                        )
+                    )
```

Comment:
I think it is possible to add a position multiple times here (if it is within multiple shapes). Shouldn't it only add  the position when the position is withing all shapes??